EIPS/eip-7400.md

```diff
@@ -0,0 +1,211 @@
+- Some lenders offer a single entry point into the protocol regardless of whether you're buying, selling, depositing or chaining them together as a flash loan, whereas other protocols offer discrete entry points.
+- Some lenders allow to flash mint any amount of their native asset without charging a fee, effectively allowing flash loans bounded by computational constraints instead of asset ownership constraints.
+
+## Specification
+
+A flash loan integrates two smart contracts using a callback pattern. These are called the `lender` and the `callback receiver` in this EIP.
+
+In the same flash loan, assets are transferred to the `loan receiver` before the callback and expected in the `payment receiver` at the end of the callback. An additional amount of assets, called a `fee`, MIGHT be also required to the present in the `payment receiver` at the end of the callback.
+
+### Lender Specification
+
+A `lender` MUST implement the IERCXXXXFlashLender interface.
+```
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.0 <0.9.0;
+import { IERC20 } from "./IERC20.sol";
+
+interface IERC3156PPFlashLender {
+
+    /// @dev The fee to be charged for a given loan.
+    /// @param asset The loan currency.
+    /// @param amount The amount of assets lent.
+    /// @return The amount of `asset` to be charged for the loan, on top of the returned principal. Returns type(uint256).max if the loan is not possible.
+    function flashFee(
+        IERC20 asset,
+        uint256 amount
+    ) external view returns (uint256);
+
+    /// @dev Use the aggregator to serve an ERC3156++ flash loan.
+    /// @param loanReceiver The address receiving the assets in the flash loan
+    /// @param asset The asset to be loaned
+    /// @param amount The amount to loaned
+    /// @param data The ABI encoded user data
+    /// @param callback The address and signature of the callback function
+    /// @return result ABI encoded result of the callback
+    function flashLoan(
+        address loanReceiver,
+        IERC20 asset,
+        uint256 amount,
+        bytes calldata data,
+        /// @dev callback. This is a concatenation of (address, bytes4), where the address is the callback receiver, and the bytes4 is the signature of callback function.
+        /// @param initiator The address that called this function
+        /// @param paymentReceiver The address that needs to receive the amount plus fee at the end of the callback
+        /// @param asset The asset to be loaned
+        /// @param amount The amount to loaned
+        /// @param fee The fee to be paid
+        /// @param data The ABI encoded data to be passed to the callback
+        /// @return result ABI encoded result of the callback
+        function(address, address, IERC20, uint256, uint256, bytes memory) external returns (bytes memory) callback
```

Comment:
function callbacks are not an officially defined type in the ABIv2 spec, so this behavior is potentially undefined and specific to how Solidity handles it internally

Reply:
Interesting, I wanted to ask you specifically about this.

I would prefer to do the specification in YAML, and I assume in that case the callback would be split into a callback selector and a callback address.

Then, I assume that in the solidity reference implementation I could use a callback type, while in a potential vyper implementation of would be to separate variables.

Would that be acceptable?

Reply:
Currently, I believe this code produces a ABI tuple `(address, bytes4)` which would be more ideal for an ERC. You could make a note that Solidity will interpret a function pointer as a tuple and vice versa, and also mention the constraints that the function should have (i.e. it returns `bytes`, it's `nonpayable`, etc.)

Reply:
I'm thinking that the function call type in solidity is `abi.encodePacked(address, selector)`, meaning that maybe vyper would need to implement the same type 🤔

Reply:
nope, nevermind it gets encoded as `bytes24` according to docs:
<img width="608" alt="image" src="https://github.com/ethereum/EIPs/assets/3859395/259da54c-9b55-45d7-ae34-2b9faefc8e83">

Reply:
https://docs.soliditylang.org/en/latest/types.html#function-types

Reply:
So yeah, I would just put this as a `callback: bytes24` type (Vyper can handle that) and make a note that `function ....` in Solidity is converted to `bytes24` packed encoding of `address` + `bytes4` selector which can be used for processing the callback.

Reply:
Sounds good to me.

However, in putting the callback as a `bytes24` I think I'll have to specify that the 4 bytes are a function selector, and that the function selector MUST include the specific list of arguments described.

A bit cumbersome, but doable.

It would be easier if vyper would include a function call type the same as solidity :D

Reply:
Perhaps, but remember this isn't standardized behavior even in solidity, and could change any time

Also, remember that the ABI spec does not specify return values in computing a selector, nor does it specify mutability

What's wrong with just defining a specific callback function name that targets should implement? There's already more specific args in the function pointer, so it's not really a generic callback, might as well just make it simple and define the interface type and make this arg a simple `ERCXXXCallbackInterface` type

Reply:
Often there are several possible flows in the flash loan callback (leverage/deleverage, borrow/repay, etc) which the user needs to define an enum for, then encode in `data`, then decode in the callback, to finally route the flow to the right spot. [Example](https://github.com/yieldprotocol/yvarb/blob/7ac3648142ff48bbd0c12e429885697001007993/src/YieldNotionalLever.sol#L274).

Allowing the callback signature to be user defined removes this, allowing the user to specify that he wants the callback to be `Foo.bar` or `Foo.baz`.

It looks complicated in the spec, but not so much in the [implementation] (https://github.com/alcueca/erc3156pp/blob/56ba9dfdee8ad34150bc77de2bb67c950a1bb415/src/FlashLender.sol#L78).

Reply:
It also allows arbitrary callback target, is that intentional?

Reply:
Yes, this standard aims to segregate the asset and control flows on flash loans.

Reply:
Some encoding and decoding of function types to bytes24 [here](https://github.com/alcueca/erc7399-wrappers/blob/ca5cad7c5387b7aa582760f1d0aa24c28771357a/src/dydx/DYDXWrapper.sol#L120). It does produce much cleaner code than using `.call`.

Reply:
Please let's continue this conversation [here](https://ethereum-magicians.org/t/erc7400-flash-loans/15211), I'd like to merge this PR and link the .md as soon as possible to avoid confusion between the PR number and the EIP number.